File: src/shared/e2ee/keyschedule.ts
/*
 * taskwraith-e2ee-v1 — handshake transcript + key schedule.
 */

import { createHash, hkdfSync, type KeyObject } from 'crypto'
import { E2EE_PROTOCOL, HKDF_INFO_IPHONE_TO_MAC, HKDF_INFO_MAC_TO_IPHONE } from './protocol'
import { deriveSharedSecret } from './keys'

export interface TranscriptInputs {
  sessionId: string
  clientEphemeralPubKeyB64: string
  serverEphemeralPubKeyB64: string
  clientNonceB64: string
  serverNonceB64: string
  /** Raw Ed25519 (b64) — binding BOTH long-lived identities into the
   * signed transcript + confirm code defeats identity-splicing: a relay
   * that swaps either identity changes the code the user compares AND
   * breaks the serverAuth signature the phone verifies. */
  macIdentityPubKeyB64: string
  iphoneIdentityPubKeyB64: string
}

/**
 * SHA-256 over the canonical transcript string. Both sides sign this (binding
 * their identity to the exact ephemerals + nonces) and derive the confirm code
 * from it — so a relay that swaps an ephemeral key produces a different hash →
 * different confirm code → the user sees a mismatch (MITM defeated).
 */
export function computeTranscriptHash(inputs: TranscriptInputs): Buffer {
  const transcript = [
    E2EE_PROTOCOL,
    inputs.sessionId,
    inputs.clientEphemeralPubKeyB64,
    inputs.serverEphemeralPubKeyB64,
    inputs.clientNonceB64,
    inputs.serverNonceB64,
    inputs.macIdentityPubKeyB64,
    inputs.iphoneIdentityPubKeyB64
  ].join('|')
  return createHash('sha256').update(transcript, 'utf8').digest()
}

/** 6-digit confirmation code from the transcript hash (shown on both screens). */
export function confirmCodeFromTranscript(transcriptHash: Buffer): string {
  return String(transcriptHash.readUInt32BE(0) % 1_000_000).padStart(6, '0')
}

export interface SessionKeys {
  /** AES-256-GCM key for the mac->iphone direction. */
  macToIphone: Buffer
  /** AES-256-GCM key for the iphone->mac direction. */
  iphoneToMac: Buffer
}

/**
 * Derive the two directional AES-256-GCM keys. `ikm` is the X25519 shared
 * secret; `salt` binds both handshake nonces; HKDF info strings keep the two
 * directions on separate keys (no nonce reuse across directions).
 */
export function deriveSessionKeys(args: {
  myEphemeralPrivate: KeyObject
  peerEphemeralPublic: KeyObject
  clientNonce: Buffer
  serverNonce: Buffer
}): SessionKeys {
  const ikm = deriveSharedSecret(args.myEphemeralPrivate, args.peerEphemeralPublic)
  const salt = createHash('sha256')
    .update(Buffer.concat([args.clientNonce, args.serverNonce]))
    .digest()
  return {
    macToIphone: Buffer.from(hkdfSync('sha256', ikm, salt, HKDF_INFO_MAC_TO_IPHONE, 32)),
    iphoneToMac: Buffer.from(hkdfSync('sha256', ikm, salt, HKDF_INFO_IPHONE_TO_MAC, 32))
  }
}

File: ios/TaskWraithKit/Sources/TaskWraithKit/KeySchedule.swift
// taskwraith-e2ee-v1 — transcript + key schedule (Swift port of
// src/shared/e2ee/keyschedule.ts).

import Foundation
import CryptoKit

public struct TranscriptInputs: Sendable {
    public var sessionId: String
    public var clientEphemeralPubKeyB64: String
    public var serverEphemeralPubKeyB64: String
    public var clientNonceB64: String
    public var serverNonceB64: String
    /// Raw Ed25519 (b64) — binding BOTH long-lived identities into the
    /// signed transcript + confirm code defeats identity-splicing (matches
    /// the Node lib's v2 transcript).
    public var macIdentityPubKeyB64: String
    public var iphoneIdentityPubKeyB64: String

    public init(
        sessionId: String, clientEphemeralPubKeyB64: String, serverEphemeralPubKeyB64: String,
        clientNonceB64: String, serverNonceB64: String,
        macIdentityPubKeyB64: String, iphoneIdentityPubKeyB64: String
    ) {
        self.sessionId = sessionId
        self.clientEphemeralPubKeyB64 = clientEphemeralPubKeyB64
        self.serverEphemeralPubKeyB64 = serverEphemeralPubKeyB64
        self.clientNonceB64 = clientNonceB64
        self.serverNonceB64 = serverNonceB64
        self.macIdentityPubKeyB64 = macIdentityPubKeyB64
        self.iphoneIdentityPubKeyB64 = iphoneIdentityPubKeyB64
    }
}

public struct SessionKeys: Sendable {
    public var macToIphone: SymmetricKey
    public var iphoneToMac: SymmetricKey
}

public enum TWKeySchedule {
    /// SHA-256 over the pipe-joined canonical transcript string.
    public static func transcriptHash(_ inputs: TranscriptInputs) -> Data {
        let transcript = [
            TWProtocol.id,
            inputs.sessionId,
            inputs.clientEphemeralPubKeyB64,
            inputs.serverEphemeralPubKeyB64,
            inputs.clientNonceB64,
            inputs.serverNonceB64,
            inputs.macIdentityPubKeyB64,
            inputs.iphoneIdentityPubKeyB64,
        ].joined(separator: "|")
        return Data(SHA256.hash(data: Data(transcript.utf8)))
    }

    /// 6-digit code from the first 4 transcript bytes (big-endian), mod 1e6.
    public static func confirmCode(_ transcriptHash: Data) -> String {
        let u32 =
            (UInt32(transcriptHash[0]) << 24) | (UInt32(transcriptHash[1]) << 16)
            | (UInt32(transcriptHash[2]) << 8) | UInt32(transcriptHash[3])
        return String(format: "%06u", u32 % 1_000_000)
    }

    /// Two directional AES-256-GCM keys. `ikm` is the raw X25519 shared secret;
    /// `salt = SHA256(clientNonce || serverNonce)`; HKDF info strings keep the
    /// directions on separate keys. Matches Node `hkdfSync('sha256', ...)`.
    public static func deriveSessionKeys(
        ikm: Data, clientNonce: Data, serverNonce: Data
    ) -> SessionKeys {
        let salt = Data(SHA256.hash(data: clientNonce + serverNonce))
        let prk = SymmetricKey(data: ikm)
        let macToIphone = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: prk, salt: salt,
            info: Data(TWProtocol.hkdfInfoMacToIphone.utf8), outputByteCount: 32)
        let iphoneToMac = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: prk, salt: salt,
            info: Data(TWProtocol.hkdfInfoIphoneToMac.utf8), outputByteCount: 32)
        return SessionKeys(macToIphone: macToIphone, iphoneToMac: iphoneToMac)
    }
}
